---
title: "Single_cell_analysis_of_unknown_cells"
author: "Thomas Sirchi"
format: html
editor: visual
---

#### Load the necessary packages to perform the analysis

```{r}
#| warning: false
#| label: load-packages
#| include: false

library(tidyverse) # collection of packages for tidy data analysis 
library(Seurat) # single cell analysis
library(patchwork) # combine separate ggplots into the same graphic
library(HGNChelper) #contains functions for identifying and correcting HGNC human gene symbols and MGI mouse gene symbols # load libraries
lapply(c("ggraph","igraph","tidyverse", "data.tree"), library, character.only = T)

```

## Analysis of Unknown cells

### Reading in the data.

```{r}
#| label: Reading
#| warning: false
XUR <- readRDS("./data/XUR.rds") 

```

### Initialize the Seurat object with the digital count matrix

Use the count matrix to create a Seurat object. The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset. Seurat v5 assays store data in layers. These layers can store raw, un-normalized counts (layer='counts'), normalized data (layer='data'), or z-scored/variance-stabilized data (layer='scale.data').

```{r}
#| label: create_seurat
#| warning: false

XUR <- CreateSeuratObject(counts = XUR, 
                           project = "XUR", # name of the project
                           min.cells = 3,   # filter for genes (rows)
                           min.features = 50 # filter for cells (columns)
                           )
XUR

```

Access the count matrix and check the data

```{r}
#| label: Reading2
#| warning: false
#XUR[["RNA"]]$counts[500:505, 1:30]
LayerData(XUR, assay = "RNA", layer = "counts")[500:505, 1:30]
```

## Standard workflow

### QC and selecting cells for further analysis

A few QC metrics commonly used by the community include:

1- The number of unique genes detected in each cell (low-quality cells or empty droplets will often have very few genes, while cell doublets or multiplets may exhibit an aberrantly high gene count)

2- The total number of molecules detected within a cell (correlates strongly with unique genes)

3- The percentage of reads that map to the mitochondrial genome (low-quality or dying cells often exhibit extensive mitochondrial contamination)

The PercentageFeatureSet function calculates the percentage of counts originating from a set of features (for example, you can use the set of all genes starting with MT- as a set of mitochondrial genes).

The number of unique genes (nFeature_RNA) and total molecules (nCount_RNA) are automatically calculated during CreateSeuratObject. You can find them stored in the object meta data.

```{r }
#| label: qc_metrics
#| warning: false
 
XUR[["percent_mt"]] <- PercentageFeatureSet(XUR, pattern = "^MT-")
XUR[["percent_ERCC"]] <- PercentageFeatureSet(XUR, pattern = "^ERCC-")
# Show QC metrics for the first 5 cells
head(XUR@meta.data, 5)
```

#### Visualize QC metrics as a violin plot

```{r }
#| label: qc_violin
#| warning: false
#| tidy: true
#| 
 
p1<- VlnPlot(XUR, features = c("nFeature_RNA", "nCount_RNA", "percent_ERCC"), ncol = 3, pt.size = 0.01)
p1
rm(p1)
```

#### FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

```{r }
#| label: qc_scatter
#| warning: false
 
plot1 <- FeatureScatter(XUR, feature1 = "nCount_RNA", feature2 = "percent_ERCC")
plot2 <- FeatureScatter(XUR, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
rm(plot1,plot2)
```

#### Filter cells based on QC values (nFeature and percent_ERCC)

```{r }
#| label: qc_filter
#| warning: false

XUR <- subset(XUR, subset = nFeature_RNA > 1500 & 
                  nFeature_RNA < 6000 & 
                  percent_ERCC < 80)
XUR
```

### Normalize data

After removing unwanted cells from the dataset, the next step is to normalize the data.

By default, Seurat employs a global-scaling normalization method "LogNormalize" that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default)(to reduce technical variation from sequencing depth), and log-transforms the result (to reduce skewness). Normalized values are stored in the data layer (XUR\[\["RNA"\]\]\$data).

Questionable assumption: each cell should have the same number of reads.

Alternative normalization methods are also available (e.g. sctransform) ps. I prefer direct code like this

```{r}
#| label: normalization
#| warning: false
XUR <- NormalizeData(XUR, 
                      normalization.method = "LogNormalize",
                     verbose = T,
                     scale.factor = 10000)

XUR[["RNA"]]$data[1:10,1:30]
```

### Identification of highly variable features (feature selection)

We next select a subset of features (genes) that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

The procedure to select variable features is implemented in the FindVariableFeatures function (the procedure models the mean-variance relationship inherent in single-cell data). By default, the function returns the 2,000 most variable features per dataset. These will be used in downstream analysis, like PCA.

```{r}
#| label: variable_features
#| warning: false

XUR <- FindVariableFeatures(XUR, 
                             selection.method = "vst", #mean.var.plot
                             nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(XUR), 10)

# plot variable features with labels
plot1 <- VariableFeaturePlot(XUR)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE, xnudge = 0, ynudge = 0)
plot2
rm(plot1,plot2,top10)
```

### Scaling the data

By scaling, Seurat applies a linear transformation to the expression levels of each gene, that is a standard pre-processing step prior to dimensional reduction techniques like PCA.

The ScaleData function:

-   Shifts the expression of each gene, so that the mean expression across cells is 0
-   Scales the expression of each gene, so that the variance across cells is 1 (z-score transformation)

This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate

The results of this are stored in XUR\[\["RNA"\]\]\$scale.data

```{r}
#| label: scaling
#| warning: false

all_genes <- rownames(XUR) # perform scaling on all genes (by default, only the top 2000 are scaled)

XUR <- ScaleData(XUR, 
                  features = all_genes)

XUR[["RNA"]]$scale.data[1:5,1:10]
```

# Cell cycle analysis

A list of cell cycle markers for the S and G2/M phases,derived from Tirosh et al, 2015, is loaded with Seurat. We can segregate this list into markers of G2/M phase and markers of S phase

```{r }
#| label: cell_cycle_marker_genes
#| warning: false
s_genes <- cc.genes$s.genes
s_genes <- str_to_title(tolower(s_genes))

g2m_genes <- cc.genes$g2m.genes
g2m_genes <- str_to_title(tolower(g2m_genes))

s_genes
g2m_genes
```

# Assign Cell-Cycle Scores

First, each cell is assigned a score, based on its expression of G2/M and S phase markers. These marker sets should be anticorrelated in their expression levels, and cells expressing neither are likely not cycling and in G1 phase.

Scores are assigned in the CellCycleScoring function, which stores S and G2/M scores in object meta data, along with the predicted classification of each cell in either G2M, S or G1 phase. CellCycleScoring can also set the identity of the Seurat object to the cell-cycle phase by passing set.ident = TRUE (the original identities are stored as old.ident). Please note that Seurat does not use the discrete classifications (G2M/G1/S) in downstream cell cycle regression. Instead, it uses the quantitative scores for G2M and S phase. However, the predicted classifications are provided in case they are of interest.

```{r }
#| label: cell_cycle_scores
#| warning: false
c_cycle <- CellCycleScoring(XUR, s.features = s_genes, g2m.features = g2m_genes, set.ident = TRUE)
s_genes %in% rownames(c_cycle)
g2m_genes %in% rownames(c_cycle)
# view cell cycle scores and phase assignments
rownames(head(c_cycle))
head(c_cycle[[]])
```

### Visualize the distribution of cell cycle markers

```{r ,warning=F,message=F}
RidgePlot(c_cycle, features = c("Hmgb2","Pcna"), ncol = 2) # very low expression 
```

Running a PCA on cell cycle genes confirms that cells separate entirely by phase

```{r }
#| label: cell_cycle_PCA
#| warning: false
c_cycle <- RunPCA(c_cycle, features = c(s_genes, g2m_genes),verbose=T)
DimPlot(c_cycle, reduction="pca")
```

# Regress out cell cycle scores during data scaling

We now attempt to subtract ('regress out') this source of heterogeneity from the data. For each gene, Seurat models the relationship between gene expression and the S and G2M cell cycle scores. The scaled residuals of this model represent a 'corrected' expression matrix, that can be used downstream for dimensional reduction.

After this regression, a PCA on the variable genes no longer returns components associated with cell cycle (long time step)

```{r}
#| label: cell_cycle_Regress
#| warning: false
c_cycle <- ScaleData(c_cycle, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(c_cycle))

c_cycle <- RunPCA(c_cycle, features = VariableFeatures(c_cycle), nfeatures.print = 10)
```

```{r}
#| label: cell_cycle_PCA2
#| warning: false
c_cycle <- RunPCA(c_cycle, features = c(s_genes, g2m_genes))

DimPlot(c_cycle)

XUR <- c_cycle
rm(s_genes,g2m_genes,c_cycle)
```

### Linear dimensional reduction (PCA)

```{r }
#| label: PCA
#| warning: false
XUR <- RunPCA(XUR, 
               features = VariableFeatures(object = XUR),
               verbose = T)

```

Seurat provides several ways of visualizing both cells and features that define the PCA, including VizDimReduction, DimPlot, and DimHeatmap. PCA results are stored in XUR\[\["pca"\]\]

```{r }
#| label: VizDimLoadings_PCA
#| warning: false
VizDimLoadings(XUR, dims = 1:2, reduction = "pca")
```

```{r, }
#| label: DimPlot_PCA
#| warning: false
DimPlot(XUR, reduction = "pca")
```

DimHeatmap allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores. Setting "cells" to a number plots the 'extreme' cells on both ends of the spectrum, which dramatically speeds plotting for large datasets.

```{r pca_heatmap, warniheatmapng=F, message=F}
#| label: DimHeatmap
#| warning: false
DimHeatmap(XUR, dims = 1:2, 
           cells = 500, 
           balanced = TRUE)

```

### Determine the 'dimensionality' of the dataset

A heuristic method to decide the number of PC to consider generates an 'Elbow plot': a ranking of principle components based on the percentage of variance explained by each one (ElbowPlot function). In this example, we can observe an 'elbow' around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs.

```{r }
#| label: ElbowPlot
#| warning: false
ElbowPlot(XUR,ndims = 50)
```

### Cluster the cells

Seurat applies a graph-based clustering approach. The distance metric which drives the clustering analysis is based on previously identified PCs. The approach to partioning the cellular distance matrix into clusters is the following: we embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'. Seurat first constructs a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters function implements this procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets.

The clusters can be found using the Idents function.

```{r }
#| label: clustering
#| warning: false
XUR <- FindNeighbors(XUR, reduction = "pca", dims = 1:50)
XUR <- FindClusters(XUR, resolution = 0.5)

head(Idents(XUR), 5)
```

### Run non-linear dimensional reduction for visualization (UMAP/tSNE)

Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.

#### Visualization with tSNE (t-Stochastic Neighbourhood Embedding)

t-SNE is a graph based, non-linear dimensionality reduction technique. It projects high dimensional data onto 2D or 3D components.

Pros - t-SNE powerfully captures the non-linearity in high dimensional datasets and is able to retain the local structures in low dimensions. This is a huge improvement over PCA. t-SNE has been used as a gold standard method for scRNA-seq data visualisation.

Cons - The way t-SNE works, it is impossible for it to preserve the global structure while performing dimension reduction. Only local structures are preserved, while the distances between groups are drastically different depending on the run. - t-SNE embeds data points onto 2 or maximum 3 dimensions only. - For huge datasets, the algorithm takes a long time to run.

```{r }
#| label: tsne
#| warning: false
XUR <- RunTSNE(XUR, dims = 1:50)
DimPlot(XUR, reduction = "tsne",label=T)
```

#### Visualization with UMAP (Uniform Manifold Approximation and Projection)

UMAP is a dimension reduction technique that can be used for visualisation similarly to t-SNE, but also for general non-linear dimension reduction.

UMAP is a relatively new dimensional reduction technique introduced by McInnes et al in 2018. (McInnes, L, Healy, J, UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction, ArXiv e-prints 1802.03426, 2018)

The algorithm is graph based and principally similar to t-SNE where it constructs a high dimensional graph representation of the data, then optimizes a low-dimensional graph to be as structurally similar as possible.

Pros - Non linear datasets: UMAP is manifold learning dimension reduction technique and thus captures the non linearity of real world datasets. It is comparable to t-SNE in terms of data visualisation. - The mathematical improvements in UMAP allow superior run time performance over t-SNE - In comparison to t-SNE, UMAP offers better preservation of a data’s global structure. - Unlike t-SNE, UMAP has no computational restrictions on embedding dimensions and can be used as an effective pre-processing step to boost the performance of density based clustering algorithms.

Cons - Lacks interpretability: Unlike PCA, where the principal components are directions of greatest variance of the source data, the lower dimension embeddings of UMAP lack strong interpretability. - One of the core assumptions of UMAP is that there exists manifold structure in the data. Because of this, UMAP can tend to find manifold structure within the noise of a dataset.

```{r }
#| label: umap
#| warning: false
XUR <- RunUMAP(XUR, dims = 1:50)
DimPlot(XUR, reduction = "umap",label=T )
```

### Finding differentially expressed features (cluster biomarkers)

Seurat can find markers that define clusters via differential expression. By default, setting only ident.1, it identifes positive and negative markers of a single cluster (specified in ident.1), compared to all other cells. FindAllMarkers automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the logfc.threshold argument requires a feature to be differentially expressed (on average) by some amount between the two groups.

The default test used is the Wilcoxon Rank Sum test

#### Find all markers of cluster 0

```{r }
#| label: markers_0
#| warning: false
c0_markers <- FindMarkers(XUR, 
                          ident.1 = 0, 
                          min.pct = 0.25) #only test genes that are detected in a minimum fraction of min.pct cells in either of the two populations
c1_markers <- FindMarkers(XUR, 
                          ident.1 = 1, 
                          min.pct = 0.25) #only test genes that are detected in a minimum fraction of min.pct cells in either of the two populations
c3_markers <- FindMarkers(XUR, 
                          ident.1 = 3, 
                          min.pct = 0.25) #only test genes that are detected in a minimum fraction of min.pct cells in either of the two populations
head(c0_markers, n = 5)
head(c1_markers, n = 5)
head(c3_markers, n = 5)
```

The results data frame has the following columns :

p_val : p_val (unadjusted)

avg_log2FC : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.

pct.1 : The percentage of cells where the feature is detected in the first group

pct.2 : The percentage of cells where the feature is detected in the second group

p_val_adj : Adjusted p-value, based on Bonferroni correction using all features in the dataset.

The following differential expression tests are currently supported:

“wilcox” : Wilcoxon rank sum test (default)

“bimod” : Likelihood-ratio test for single cell feature expression, (McDavid et al., Bioinformatics, 2013)

“roc” : Standard AUC classifier. For each gene, evaluates (using AUC) a classifier built on that gene alone, to classify between two groups of cells. An AUC value of 1 means that expression values for this gene alone can perfectly classify the two groupings (i.e. Each of the cells in cells.1 exhibit a higher level than each of the cells in cells.2). An AUC value of 0 also means there is perfect classification, but in the other direction. A value of 0.5 implies that the gene has no predictive power to classify the two groups. Returns a 'predictive power' (abs(AUC-0.5) \* 2) ranked matrix of putative differentially expressed genes.

“t” : Student’s t-test

“poisson” : Likelihood ratio test assuming an underlying poisson distribution. Use only for UMI-based datasets

“negbinom” : Likelihood ratio test assuming an underlying negative binomial distribution. Use only for UMI-based datasets

“LR” : Uses a logistic regression framework to determine differentially expressed genes. Constructs a logistic regression model predicting group membership based on each feature individually and compares this to a null model with a likelihood ratio test.

“MAST” : Utilizes the MAST package to run the DE testing. GLM-framework that treates cellular detection rate as a covariate (Finak et al, Genome Biology, 2015)

“DESeq2” : Identifies differentially expressed genes between two groups of cells based on a model using DESeq2 which uses a negative binomial distribution (Love et al, Genome Biology, 2014)

#### Find all markers distinguishing cluster 0 from clusters 1, 2 and 3

```{r}
#| label: markers_3
#| warning: false
c0_1_3_markers <- FindMarkers(XUR, 
                          ident.1 = 0, 
                          ident.2 = c(1,3), 
                          min.pct = 0.25)
head(c3_markers, n = 5)
```

#### Find markers for every cluster compared to all remaining cells, report only the positive ones

```{r }
#| label: all_markers
#| warning: false
XUR_markers <- FindAllMarkers(XUR, 
                               only.pos = TRUE, 
                               min.pct = 0.25, 
                               logfc.threshold = 0.25)

XUR_markers %>%
    group_by(cluster) %>%
    slice_max(n=2,order_by=avg_log2FC) 
```

### Visualization tools

Seurat offers several tools for visualizing marker expression.

#### VlnPlot shows expression probability distributions across clusters

```{r }
#| label: vlnplot
#| warning: false
VlnPlot(XUR, features = c("1500015O10Rik", "Spns3"),pt.size=0)
```

#RidgePlot also shows expression probability distributions

```{r }
#| label: ridgeplot
#| warning: false
RidgePlot(XUR, features = c("1500015O10Rik"))
```

# FeaturePlot visualizes feature expression on a tSNE, UMAP or PCA plot

```{r }
#| label: featureplot
#| warning: false
FeaturePlot(XUR, features = c("1500015O10Rik", "Spns3"), order =T )
```

# DotPlot

Intuitive way of visualizing how feature expression changes across different identity classes (clusters). The size of the dot encodes the percentage of cells within a class, while the color encodes the AverageExpression level of cells within a class (blue is high).

```{r dotplot, warning=F, message=F}
#| label: dotplot
#| warning: false
#| tidy: true
DotPlot(XUR, features = c("1500015O10Rik", "Chac1", "Spns3","Foxj1"))
```

#### DoHeatmap generates an expression heatmap for given cells and features. In this case, we are plotting the top 3 markers (or all markers if less than 3) for each cluster.

```{r }
#| label: geatmap
#| warning: false
#| tidy: true
top3 <- XUR_markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_log2FC)
DoHeatmap(XUR, features = top3$gene) + NoLegend()
rm(top3)
```

# Interactive plotting features

Seurat utilizes R’s plotly graphing library to create interactive plots.

```{r}
#| label: Interactive plotting features
#| warning: false
#| tidy: true
#plot <- FeaturePlot(XUR, features = "1500015O10Rik")
#HoverLocator(plot = plot, information = FetchData(XUR, vars = c("ident", "nFeature_RNA")))
#rm(plot)
```

# Marker-based automatic annotation

Marker-based automatic annotation labels cells or cell clusters based on the characteristic expression of known marker genes. Set of marker genes can be obtained from databases (PanglaoDB, CellMarker, CellTypist ...) or manually from the literature.

To be successful, the marker gene or gene set (a collection of marker genes) should be specifically and consistently expressed in a given cell, cluster or class of cells. Marker-based automatic annotation works well once a relevant and sufficiently large set of marker genes is collected

# scType (Ianevski et al. 2022)

ScType accepts both positive and negative markers, i.e., gene that are not expected to be expressed in a particular cell type. Sctype provides its own marker database for human and mouse, obtained from the integration of the information available in the CellMarker database (<http://biocc.hrbmu.edu.cn/CellMarker/>) and PanglaoDB (<https://panglaodb.se>). In total, the current version of the ScType database comprises 3,980 cell markers for 194 cell types in 17 human tissues and 4,212 cell markers for 194 cell types in 17 mouse tissues.

scType cell_type annotation: 1. For each positive/negative marker compute specificity score, which indicate whether a gene is a marker for a specific cell types. 2. The raw expression matrix is normalized and Z-transform (scale the expression of each gene across cells) 3. The transformed matrix is multiply by the cell-type specificity score 4. For each cell types the expression scores of all its positive markers are summarized into a single enrichment score by summing them and dividing by square root of their number. The same is done for the negative markers. 5. The negative marker expression score is subtracted from the positive score to obtain the final enrichment score. Individual cells are assigned to a cell type based on the maximum value for the cell type marker set.

#Automatically detect a tissue type of the dataset (did it manualy)

```{r }
#| label: Find_tissue
#| warning: false
#| tidy: true
# DB file
db_ <- "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_full.xlsx"
#db_ <- "ScTypeDB_full.xlsx"
tissue <- c("Brain","Immune system","Pancreas","Liver","Eye","Kidney","Brain","Lung","Adrenal","Heart","Intestine","Muscle","Placenta","Spleen","Stomach","Thymus")
tissue

# Load gene set preparation function
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R")

# Load cell type annotation function
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")

# Initialize a list to store sctype_scores for each tissue
sctype_scores_list <- list()
score_sum_list <- list()  # Initialize a list to store sum of scores for each tissue

# Loop over each tissue
for (t in tissue) {
  cat("Now analyzing for tissue", t, "...\n")  # Print current tissue being analyzed
  # Prepare gene sets for the current tissue
  gs_list <- gene_sets_prepare(db_, t)
  
  # Calculate ScType scores for the current tissue
  es.max <- sctype_score(scRNAseqData = XUR[["RNA"]]$scale.data, scaled = TRUE, gs = gs_list$gs_positive, gs2 = NULL) #gs_list$gs_negative
  
  # Merge by cluster
  cL_resutls <- do.call("rbind", lapply(unique(XUR@meta.data$seurat_clusters), function(cl){
    es.max.cl = sort(rowSums(es.max[, rownames(XUR@meta.data[XUR@meta.data$seurat_clusters==cl, ])]), decreasing = TRUE)
    head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(XUR@meta.data$seurat_clusters==cl)), 10)
  }))
  sctype_scores <- cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  
  
  # Set low-confident (low ScType score) clusters to "unknown"
  sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] <- "Unknown"
  # Store sctype_scores in the list
  sctype_scores_list[[t]] <- sctype_scores[, 1:3]
  
  # Calculate sum of scores for the current tissue and store it in the score_sum_list
  score_sum_list[[t]] <- sum(sctype_scores$scores)
  #XUR[["scType_labels"]] <- sctype_scores$type[match(XUR@meta.data$seurat_clusters, sctype_scores$cluster)]
  #DimPlot(XUR, reduction = "umap", label = TRUE, repel = TRUE, group.by = 'scType_labels') 
}

# Combine tissue names and corresponding score sums into a data frame
score_sum_df <- data.frame(tissue = names(score_sum_list), score_sum = unlist(score_sum_list))
score_sum_df<- score_sum_df %>% 
  arrange(score_sum)
score_sum_df

# Create a bar plot using ggplot2 with the ordered dataframe
ggplot(score_sum_df, aes(x = reorder(tissue, score_sum), y = score_sum, fill = tissue)) +
  geom_bar(stat = "identity", color = "black",show.legend = FALSE) +
  #scale_fill_brewer(palette = "RdPu",type = "seq") +  # Set the fill color scale
  labs(title = "Sum of ScType Scores for Each Tissue",
       x = "Tissue", y = "Sum of Scores") +
  theme_minimal() +
  coord_flip()  # Flip the plot
```

# Assign cell types to each cluster with Wrapper function

```{r }
#| label: wrapper_function
#| warning: false
#| tidy: true
# get cell-type by cell matrix
source("https://raw.githubusercontent.com/kris-nader/sc-type/master/R/sctype_wrapper.R")
XUR <- run_sctype(XUR,known_tissue_type="Brain",custom_marker_file =db_ ,name="scType_labels",plot=TRUE)
```

```{r}
#| label: booble_plot
#| warning: false
#| tidy: true
# prepare edges
cL_resutls <- cL_resutls[order(cL_resutls$cluster),]; edges = cL_resutls; edges$type = paste0(edges$type,"_",edges$cluster); edges$cluster = paste0("cluster ", edges$cluster); edges = edges[,c("cluster", "type")]; colnames(edges) = c("from", "to"); rownames(edges) <- NULL

# prepare nodes
nodes_lvl1 <- sctype_scores[,c("cluster", "ncells")]; nodes_lvl1$cluster = paste0("cluster ", nodes_lvl1$cluster); nodes_lvl1$Colour = "#f1f1ef"; nodes_lvl1$ord = 1; nodes_lvl1$realname = nodes_lvl1$cluster; nodes_lvl1 = as.data.frame(nodes_lvl1); nodes_lvl2 = c(); 
ccolss <- c("#5f75ae","#92bbb8","#64a841","#e5486e","#de8e06","#eccf5a","#b5aa0f","#e4b680","#7ba39d","#b15928","#ffff99", "#6a3d9a","#cab2d6","#ff7f00","#fdbf6f","#e31a1c","#fb9a99","#33a02c","#b2df8a","#1f78b4","#a6cee3")
for (i in 1:length(unique(cL_resutls$cluster))){
  dt_tmp = cL_resutls[cL_resutls$cluster == unique(cL_resutls$cluster)[i], ]; nodes_lvl2 = rbind(nodes_lvl2, data.frame(cluster = paste0(dt_tmp$type,"_",dt_tmp$cluster), ncells = dt_tmp$scores, Colour = ccolss[i], ord = 2, realname = dt_tmp$type))
}
nodes <- rbind(nodes_lvl1, nodes_lvl2); nodes$ncells[nodes$ncells<1] = 1;
files_db <- openxlsx::read.xlsx(db_)[,c("cellName","shortName")]; files_db = unique(files_db); nodes = merge(nodes, files_db, all.x = T, all.y = F, by.x = "realname", by.y = "cellName", sort = F)
nodes$shortName[is.na(nodes$shortName)] = nodes$realname[is.na(nodes$shortName)]; nodes = nodes[,c("cluster", "ncells", "Colour", "ord", "shortName", "realname")]

mygraph <- graph_from_data_frame(edges, vertices=nodes)

# Make the graph
gggr <- ggraph(mygraph, layout = 'circlepack', weight=I(ncells)) + 
  geom_node_circle(aes(filter=ord==1,fill=I("#F5F5F5"), colour=I("#D3D3D3")), alpha=0.9) + geom_node_circle(aes(filter=ord==2,fill=I(Colour), colour=I("#D3D3D3")), alpha=0.9) +
  theme_void() + geom_node_text(aes(filter=ord==2, label=shortName, colour=I("#ffffff"), fill="white", repel = !1, parse = T, size = I(log(ncells,25)*1.5)))+ geom_node_label(aes(filter=ord==1,  label=shortName, colour=I("#000000"), size = I(3), fill="white", parse = T), repel = !0, segment.linetype="dotted")
  
DimPlot(XUR, reduction = "umap", label = TRUE, repel = TRUE, cols = ccolss)+ gggr
```
